subsys/net/lib/tls_credentials/tls_credentials_trusted.c

```diff
@@ -456,3 +460,399 @@ int tls_credential_delete(sec_tag_t tag, enum tls_credential_type type)
 
 	return ret;
 }
+
+#if defined(CONFIG_TLS_CREDENTIAL_KEYGEN)
+
+#include <psa/crypto.h>
+#include <mbedtls/pk.h>
+#include <mbedtls/asn1.h>
+
+/* For this backend, the private-key will be stored in DER format.
+ * Use `cred get <tag> PK bin` to retrieve (the non-terminated base64 encoding of) the key using
+ * the credential shell.
+ */
+int tls_credential_keygen(sec_tag_t tag, enum tls_credential_keygen_type type,
+			  void *key_buf, size_t *key_len)
+{
+	int ret = 0;
+
+	/* PSA keygen */
+	psa_status_t psa_status;
+	psa_key_attributes_t key_attributes = PSA_KEY_ATTRIBUTES_INIT;
+	psa_key_id_t key_id = PSA_KEY_ID_NULL;
+	psa_key_usage_t key_usage;
+
+	/* MbedTLS key formatting */
+	int mbed_status;
+	mbedtls_pk_context pk_ctx;
+	void* key_start = 0;
+	size_t key_max;
+
+	/* Permanent key storage */
+	int cred_status;
+	psa_storage_uid_t uid = tls_credential_get_uid(tag, TLS_CREDENTIAL_PRIVATE_KEY);
+	bool key_stored = false;
+
+	if (key_len == NULL) {
+		return -EINVAL;
+	}
+
+	key_max = *key_len;
+
+	k_mutex_lock(&credential_lock, K_FOREVER);
+	mbedtls_pk_init(&pk_ctx);
+
+	/* Ensure the storage destination is valid. */
+	if (!tag_type_valid(tag, TLS_CREDENTIAL_PRIVATE_KEY)) {
+		ret = -EINVAL;
+		goto cleanup;
+	}
+
+	/* Presently, only SECP256R1 is supported by this backend for keygen */
```

Comment:
Well this is an "artificial" limitation to me that you can easily lift. In `tls_credentials.h` you are already defining a subset of known EC keys in `enum tls_credential_keygen_type`, but then you only implement one type here... It would be nice to either:

- only include the curve types really supported in `enum tls_credential_keygen_type` or,
- extend the support on lines 544-545 in order to support all the groups.

I prefer the 2nd option.

Reply:
Since this is a generic API, `enum tls_credential_keygen_type` needs to be comprehensive enough to cover all possible/conceivable/reasonable APIs, hence why I included (nearly) "full set" of reasonable values on the enum as a starting point.

Support for more key types in this backend would definitely be a good idea, but for this PR I would like to limit it to just `SECP256R1`, since that is what my team needs ATM, and to support more would require significantly more testing work. Support for additional key types can be added in the future I think